File: android/app/src/main/java/com/carelog/ui/vitals/TemperatureViewModel.kt
package com.carelog.ui.vitals

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.carelog.auth.AuthRepository
import com.carelog.fhir.client.ObservationType
import com.carelog.fhir.models.FhirObservation
import com.carelog.fhir.repository.LocalFhirRepository
import com.carelog.voice.VoiceAcknowledgementPlayer
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.Instant
import java.util.UUID
import javax.inject.Inject

/**
 * ViewModel for temperature logging screen.
 */
@HiltViewModel
class TemperatureViewModel @Inject constructor(
    private val authRepository: AuthRepository,
    private val localFhirRepository: LocalFhirRepository,
    private val voicePlayer: VoiceAcknowledgementPlayer
) : ViewModel() {

    private val _uiState = MutableStateFlow(TemperatureUiState())
    val uiState: StateFlow<TemperatureUiState> = _uiState.asStateFlow()

    fun updateValue(value: String) {
        _uiState.update { state ->
            val doubleValue = value.toDoubleOrNull()
            val error = validateValue(doubleValue, state.unit)
            state.copy(
                value = value,
                valueError = error,
                canSave = error == null && doubleValue != null
            )
        }
    }

    fun updateUnit(unit: String) {
        _uiState.update { state ->
            // Convert value if switching units
            val currentValue = state.value.toDoubleOrNull()
            val newValue = if (currentValue != null) {
                if (unit == "\u00B0C" && state.unit == "\u00B0F") {
                    // Convert F to C
                    String.format("%.1f", (currentValue - 32) * 5 / 9)
                } else if (unit == "\u00B0F" && state.unit == "\u00B0C") {
                    // Convert C to F
                    String.format("%.1f", currentValue * 9 / 5 + 32)
                } else {
                    state.value
                }
            } else {
                ""
            }

            val doubleValue = newValue.toDoubleOrNull()
            val error = validateValue(doubleValue, unit)

            state.copy(
                unit = unit,
                value = newValue,
                valueError = error,
                canSave = error == null && doubleValue != null
            )
        }
    }

    private fun validateValue(value: Double?, unit: String): String? {
        if (value == null) return null

        return when (unit) {
            "\u00B0F" -> when {
                value < 90 -> "Too low (min 90\u00B0F)"
                value > 110 -> "Too high (max 110\u00B0F)"
                else -> null
            }
            "\u00B0C" -> when {
                value < 32 -> "Too low (min 32\u00B0C)"
                value > 43 -> "Too high (max 43\u00B0C)"
                else -> null
            }
            else -> null
        }
    }

    fun saveReading(onSuccess: () -> Unit) {
        val state = _uiState.value
        val value = state.value.toDoubleOrNull() ?: return

        viewModelScope.launch {
            _uiState.update { it.copy(isSaving = true, saveError = null) }

            try {
                val user = authRepository.getCurrentUser()
                val patientId = user?.linkedPatientId ?: throw IllegalStateException("No patient ID")

                // Store in Celsius (standard FHIR unit)
                val valueInCelsius = if (state.unit == "\u00B0F") {
                    (value - 32) * 5 / 9
                } else {
                    value
                }

                // Create temperature observation
                val observation = FhirObservation(
                    id = UUID.randomUUID().toString(),
                    patientId = patientId,
                    type = ObservationType.BODY_TEMPERATURE,
                    value = valueInCelsius,
                    unit = "Cel",
                    effectiveDateTime = Instant.now().toString(),
                    performerId = user?.userId
                )

                // Save to local store (adds to sync queue)
                localFhirRepository.saveObservation(observation)

                // Play voice acknowledgement
                voicePlayer.playSuccess(ObservationType.BODY_TEMPERATURE)

                _uiState.update { it.copy(isSaving = false) }
                onSuccess()

            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        isSaving = false,
                        saveError = e.message ?: "Failed to save"
                    )
                }
                voicePlayer.playFailure()
            }
        }
    }
}

/**
 * UI state for temperature screen.
 */
data class TemperatureUiState(
    val value: String = "",
    val unit: String = "\u00B0F",
    val valueError: String? = null,
    val canSave: Boolean = false,
    val isSaving: Boolean = false,
    val saveError: String? = null
)

File: ios/CareLog/CareLog/Vitals/ViewModels/TemperatureViewModel.swift
import Foundation
import SwiftUI
import Combine

/// ViewModel for temperature logging screen.
@MainActor
final class TemperatureViewModel: ObservableObject {
    // MARK: - Published Properties

    @Published var value: String = ""
    @Published var unit: String = "\u{00B0}F"
    @Published var valueError: String?
    @Published var isSaving: Bool = false
    @Published var saveError: String?

    // MARK: - Computed Properties

    var canSave: Bool {
        guard let doubleValue = Double(value) else { return false }
        return validateValue(doubleValue) == nil
    }

    // MARK: - Dependencies

    private let authService: AuthService
    private let localFHIRRepository: LocalFHIRRepository
    private let voicePlayer: VoiceAcknowledgementPlayer
    private var cancellables = Set<AnyCancellable>()

    // MARK: - Initialization

    init(
        authService: AuthService = .shared,
        localFHIRRepository: LocalFHIRRepository = .shared,
        voicePlayer: VoiceAcknowledgementPlayer = .shared
    ) {
        self.authService = authService
        self.localFHIRRepository = localFHIRRepository
        self.voicePlayer = voicePlayer

        setupValidation()
    }

    // MARK: - Validation

    private func setupValidation() {
        $value
            .dropFirst()
            .sink { [weak self] value in
                guard let self = self,
                      let doubleValue = Double(value) else {
                    self?.valueError = nil
                    return
                }
                self.valueError = self.validateValue(doubleValue)
            }
            .store(in: &cancellables)
    }

    private func validateValue(_ value: Double) -> String? {
        switch unit {
        case "\u{00B0}F":
            switch value {
            case ..<90: return "Too low (min 90\u{00B0}F)"
            case 111...: return "Too high (max 110\u{00B0}F)"
            default: return nil
            }
        case "\u{00B0}C":
            switch value {
            case ..<32: return "Too low (min 32\u{00B0}C)"
            case 44...: return "Too high (max 43\u{00B0}C)"
            default: return nil
            }
        default:
            return nil
        }
    }

    // MARK: - Unit Conversion

    func convertUnit(to newUnit: String) {
        guard let currentValue = Double(value) else { return }

        let convertedValue: Double
        if newUnit == "\u{00B0}C" && unit == "\u{00B0}F" {
            // Convert F to C
            convertedValue = (currentValue - 32) * 5 / 9
            value = String(format: "%.1f", convertedValue)
        } else if newUnit == "\u{00B0}F" && unit == "\u{00B0}C" {
            // Convert C to F
            convertedValue = currentValue * 9 / 5 + 32
            value = String(format: "%.1f", convertedValue)
        }

        // Validate with new unit
        if let doubleValue = Double(value) {
            valueError = validateValue(doubleValue)
        }
    }

    // MARK: - Save

    func saveReading() async {
        guard let tempValue = Double(value) else { return }

        isSaving = true
        saveError = nil

        do {
            guard let user = await authService.getCurrentUser(),
                  let patientId = user.linkedPatientId else {
                throw VitalError.noPatientId
            }

            // Store in Celsius (standard FHIR unit)
            let valueInCelsius = unit == "\u{00B0}F" ? (tempValue - 32) * 5 / 9 : tempValue

            // Create temperature observation
            let observation = FHIRObservation(
                patientId: patientId,
                type: .bodyTemperature,
                effectiveDateTime: Date(),
                value: valueInCelsius,
                unit: "Cel",
                performerId: user.userId,
                performerType: .relatedPerson
            )

            // Save to local store (adds to sync queue)
            try await localFHIRRepository.saveObservation(observation)

            // Play voice acknowledgement
            voicePlayer.playSuccess(type: .bodyTemperature)

            isSaving = false

        } catch {
            isSaving = false
            saveError = error.localizedDescription
            voicePlayer.playFailure()
        }
    }
}
